### PR TITLE
Nearest Rotation SVD Method

### LINKED ISSUE #1023: Nearest rotation SVD method
Several of the common-line algorithms as well as the method for registering estimated rotations to ground truth use an SVD method for finding the nearest rotations for a set of 3x3 matrices. We should add a utility function to handle these cases.

In an initial search for all uses of this method I noticed the methods for registering rotations is duplicated in utils/coor_trans.py and utils/rotation.py. This would be a good time to clean that up.

### PR DESCRIPTION
**Pull request by j-c-c**
This PR adds a utility function to find the set of nearest rotations for a 3x3 or nx3x3 array.

In the case of the nearest orthogonal matrix being a reflection, this method negates the third singular value to produce a rotation.

Closes #1023

### CODE REVIEW

File: src/aspire/utils/matrix.py [hunk 1]
```diff
@@ -434,6 +434,34 @@ def best_rank1_approximation(A):
...
+        raise ValueError(
+            f"Array must be of shape (3, 3) or (n, 3, 3). Found shape {A.shape}."
+        )
+
+    # For the singular value decomposition A = U @ S @ V, we compute the nearest rotation
+    # matrices R = U @ V. If det(U)*det(V) = -1, we negate the third singular value to ensure
+    # we have a rotation.
+    U, _, V = np.linalg.svd(A)
+    neg_det_idx = np.linalg.det(U) * np.linalg.det(V) < 0
+    U[neg_det_idx] = U[neg_det_idx] @ np.diag((1, 1, -1)).astype(dtype, copy=False)
```
Comment (garrettwrong) [inline src/aspire/utils/matrix.py:459]: I think you can just call diag(..., dtype=dtype) here

File: src/aspire/utils/matrix.py [hunk 2]
```diff
@@ -434,6 +434,34 @@ def best_rank1_approximation(A):
...
+    rots = np.einsum("ijk, ikl -> ijl", U, V)
```
Comment (garrettwrong) [inline src/aspire/utils/matrix.py]: Question, is this a really complicated way of writing U@V ?

File: tests/test_matrix.py [hunk 3]
```diff
@@ -342,3 +348,80 @@ def testFixSigns(self):
...
+    refl = rot @ np.diag((1, -1, 1)).astype(dtype)
+    noise = 1e-3 * randn(9, seed=0).astype(dtype, copy=False).reshape(3, 3)
+    noisy_refl = refl + noise
+
+    # Find nearest rotation.
+    nearest_rot = nearest_rotations(noisy_refl)
+
+    # Check that estimate is a rotation.
+    _is_rotation(nearest_rot, dtype)
+
```
Comment (garrettwrong) [inline tests/test_matrix.py:392]: Looks like forgot to compare with a known rotation.

File: tests/test_matrix.py [hunk 4]
```diff
@@ -342,3 +348,80 @@ def testFixSigns(self):
...
+    # Check that we retain singleton shape.
+    assert nearest_rot.shape == rot.shape
+
+
+def test_nearest_rotations_error():
+    # Check error for bad ndim.
+    A = randn(2 * 5 * 3 * 3, seed=0).reshape(2, 5, 3, 3)
```
Comment (garrettwrong) [inline tests/test_matrix.py]: Think you can use `np.empty` here.  For this tiny array its not a problem though.  FWIW, generating really large random arrays can actually consume some compute time...  

image

File: tests/test_matrix.py [hunk 5]
```diff
@@ -342,3 +348,80 @@ def testFixSigns(self):
...
+
+    :param R: Singleton or stack of 3x3 arrays.
+    :param dtype: dtype to use for test tolerance.
+    :return: boolean indicating if all 3x3 arrays are rotations.
+    """
+    if R.ndim == 2:
+        R = R[np.newaxis]
+
+    n_rots = len(R)
+    RTR = np.einsum("ikj,ikl->ijl", R, R)
```
Comment (garrettwrong) [inline tests/test_matrix.py]: Lets not add more `einsum` if we don't need to :).  Have low confidence most people that write them understand them later. 

I believe the guidance that `einsum` is faster than the standard built in functions is dated from 5-10 years ago and is unlikely true anymore outside a few exceptional cases.  I doubt small`gemm` calls is one of the exceptions 😇 .

I take it this was meant to be R.T @ R 😂 ?

File: tests/test_matrix.py [hunk 6]
```diff
@@ -342,3 +348,80 @@ def testFixSigns(self):
...
+    # Check error for bad shape.
+    A = randn(5 * 3 * 2, seed=0).reshape(5, 3, 2)
+    with pytest.raises(ValueError, match="Array must be of shape"):
+        _ = nearest_rotations(A)
+
+
+def _is_rotation(R, dtype):
+    """
+    Helper function to check if a set of 3x3 matrices are rotations
+    by checking that R.T @ R = I and det(R) = 1.
```
Comment (garrettwrong) [inline tests/test_matrix.py:412]: the only thing better than an orthogonal matrix is a special orthogonal matrix 😂

**Review by garrettwrong — DISMISSED** [review]
This looks really great, thanks!  A couple trivial changes, mostly just remarks.

My only other thought is whether this should just speak `Rotation` objects, but I see that it isn't used that way... so that probably would needlessly complicate things and introduce error. Likely better to avoid for now.

File: src/aspire/utils/matrix.py (hunk 1 above)
Comment (j-c-c) [reply src/aspire/utils/matrix.py:459]: I thought so too, but `diag` doesn't accept a dtype kwarg. It just mutates an array so passes the same dtype. I could do something like `np.diag(np.array([1, 1, -1], dtype=dtype))` if that's preferable.

File: src/aspire/utils/matrix.py (hunk 2 above)
Comment (j-c-c) [reply src/aspire/utils/matrix.py]: lol. yeah. Not sure why I was complicating that! Changed.

File: src/aspire/utils/matrix.py (hunk 1 above)
Comment (garrettwrong) [reply src/aspire/utils/matrix.py:459]: Yeah, you are right.  Nah lets leave it, my mistake!

File: src/aspire/utils/matrix.py (hunk 2 above)
Comment (garrettwrong) [reply src/aspire/utils/matrix.py]: I just assumed someone else wrote it 🙈 . Cool thanks

File: tests/test_matrix.py (hunk 3 above)
Comment (j-c-c) [reply tests/test_matrix.py:392]: I don't actually expect the `nearest_rot` to be very close to some known rotation in this case (ie. `refl` in the test). I know that sounds silly for a method called `nearest_rotations`, but this is just exercising the case where we're finding the nearest rotation to a rotation estimate so noisy that it's actually a reflection, ie. has negative determinant. This still happens to be the nearest rotation to that reflection (in a matrix norm sense), but giving some angular distance won't really make sense. I think the important thing to validate in this case is that we are actually getting a rotation. My guess is this won't come up in practice very often.

Comment (garrettwrong) [reply tests/test_matrix.py:392]: I just presumed by the construction you would know a rotation if the reflection was picked a certain way.

No worries, sounds more difficult than I thought. Thanks!

File: tests/test_matrix.py (hunk 5 above)
Comment (j-c-c) [reply tests/test_matrix.py]: 🤣 Sorry, was on some kind of einsum kick in this PR. Changed to `np.transpose(R, axes=(0,2,1)) @ R`.

File: tests/test_matrix.py (hunk 4 above)
Comment (j-c-c) [reply tests/test_matrix.py]: Good call. Changed this. No point computing values that will never be used.

File: tests/test_matrix.py (hunk 6 above)
Comment (j-c-c) [reply tests/test_matrix.py:412]: 😂

File: src/aspire/utils/matrix.py (hunk 2 above)
Comment (j-c-c) [reply src/aspire/utils/matrix.py]: Yeah, this one was lifted. But should have known better.

**Review by garrettwrong — APPROVED** [review]

**Comment by garrettwrong** [comment]
Restarted the one failed job, was another FLE failing to converge.

**Comment by garrettwrong** [comment]
Because the function in the SDP code shouldn't have changed, I'm fine with merging this in if you are.

**Pull request merged**